### PR TITLE
Fix another bug

### PR DESCRIPTION
**Pull request by jxmorris12**
Our metric computations were using the logits, not probabilities, in my previous pull request (sorry!). I think it's better to have the model return a probability distribution (i.e. `torch.sigmoid` on the outputs) and then handle everything as probabilities and not logits. So I had to change the loss function too.

An important note is that we've only set up fine-tuning for binary classification tasks like `rotten_tomatoes`! So if we want to fine-tune and test on a task that has >2 labels, we'll need to change the code a bit.

### CODE REVIEW

File: .github/workflows/test.yml [hunk 1]
```diff
@@ -15,7 +15,7 @@ jobs:
     runs-on: ubuntu-latest
     strategy:
       matrix:
-        python-version: [3.8]
+        python-version: [3.9]
```
Comment (jscuds) [inline .github/workflows/test.yml:18]: See [Lines 71 in `experiment.py`](https://github.com/jscuds/rf-bert/pull/12/files#diff-b2f3f93dba319f5dec38bff9562c88bdf781e8c5134de6d45fcd75542c976b87R71) && [Line 126-129 in `train.py`](https://github.com/jscuds/rf-bert/pull/12/files#diff-ed183d67207df065a11e1289f19d34cc2abbc5448dea952683cfe9728c342b95R126) (same comment for both places:)

Any issue if I change this to a method friendly to versions of Python 3.5 - 3.8?  I can explain my cluster concerns later if necessary, but right the modules that I can load for the cluster don't seem to go above Python 3.8.

`z = {**x, **y}` in Python 3.5+ is equivalent to  `z = x | y` in Python 3.9+ [StackOverflow Ref](https://stackoverflow.com/questions/38987/how-do-i-merge-two-dictionaries-in-a-single-expression-take-union-of-dictionari?page=1&tab=votes#tab-top)

File: experiment.py [hunk 2]
```diff
@@ -49,16 +49,28 @@ def compute_and_reset_metrics(self, epoch: int) -> Dict[str, float]:
...
+        # Add metrics that can't be computed by simple averaging
+        # (accuracy, precision, f1...)
+        for metric in self.metrics:
+            # Compute metrics like test/accuracy and train/accuracy
+            # and reset count.
+            metric_dict = metric.compute()
+            # Print metrics and add to list of total metrics.
+            for name, val in metric_dict.items():
+                logger.info('\t%s = %f', name, val)
+            all_metrics_dict = all_metrics_dict | metric_dict
```
Comment (jscuds) [inline experiment.py:71]: Ref comment in [test.yml](https://github.com/jscuds/rf-bert/pull/12/files#diff-faff1af3d8ff408964a57b2e475f69a6b7c7b71c9978cccc8f471798caac2c88R18)

I think this case could be `all_metrics_dict.update(metric_dict)`

File: train.py [hunk 3]
```diff
@@ -123,13 +123,17 @@ def run_training_loop(args: argparse.Namespace) -> str:
     day = time.strftime(f'%Y-%m-%d-%H%M')
     exp_name = f'{args.experiment}_{args.model_name}_{day}'
     # WandB init and config (based on argument dictionaries in imports/globals cell)
+    config_dict = vars(args) | {
```
Comment (jscuds) [inline train.py:126]: Ref. comment in [test.yml](https://github.com/jscuds/rf-bert/pull/12/files#diff-faff1af3d8ff408964a57b2e475f69a6b7c7b71c9978cccc8f471798caac2c88R18)

File: utils.py [hunk 4]
```diff
@@ -39,10 +39,17 @@ def clear_all(self) -> None:
...
-                title=f"{description} Histogram, Epoch {epoch+1}")})
+    histogram = wandb.plot.histogram(
+        table, lower_description,
+        title=f"{description} Histogram, Epoch {epoch+1}"
+    )
+    wandb.log({
+            lower_description: histogram, 
+            epoch: epoch
+            }, step=step
+    )
```
Comment (jscuds) [inline utils.py:55]: I'll have to look at this in practice (it took me a few tries to get it right the first time I wrote this).

File: experiment.py [hunk 5]
```diff
@@ -217,12 +224,14 @@ class FinetuneExperiment(Experiment):
...
     model: ElmoClassifier
-    metrics: Dict[str, Metric]
+    metrics: List[Metric]
     metric_averages: TensorRunningAverages
 
     def __init__(self, args: argparse.Namespace):
         assert args.model_name in {"elmo_single_sentence", "elmo_sentence_pair"} # TODO: Support choice of model via argparse.
         self.args = args
+
+        # TODO: how to handle req_grad_elmo args during fine-tuning? Shouldn't ELMO never be frozen?
```
Comment (jscuds) [inline experiment.py:234]: We can leave this TODO, but I agree, ELMo should _never_ be frozen during fine-tuning.

File: experiment.py [hunk 6]
```diff
@@ -261,16 +270,20 @@ def get_dataloaders(self) -> Tuple[DataLoader, DataLoader]:
...
         assert preds.shape == targets.shape
         loss = self._loss_fn(preds, targets)
         self.metric_averages.update(f'{metrics_key}/Loss', loss.item())
         pred_classes = torch.sigmoid(preds.squeeze()).detach() # TODO should we round here for sure?
 
-        for metric_name, metric_func in self.metrics.items():
-            val = metric_func(pred_classes, targets)
-            self.metric_averages.update(f'{metrics_key}/{metric_name}', val)
+        for metric in self.metrics:
+            metric.update(metrics_key, preds, targets)
```
Comment (jscuds) [inline experiment.py:284]: Just to make sure I understand this correctly:
- `self.metrics` is a list of two classes defined in `metrics.py`. 
- by using the class method `.update()` it updates the respective `Accuracy` or `PrecisionRecallF1` class which both act as a dictionary to hold the accumulating values to be computed at the end of an epoch

...right?

File: metrics.py [hunk 7]
```diff
@@ -1,27 +1,89 @@
...
+
+glob_n=0
+class Accuracy(Metric):
+    """The number of predictions for y_true that are 100% correct."""
+    num_correct: collections.defaultdict
+    total: collections.defaultdict
+    total: int
+    def __init__(self):
+        self.num_correct = collections.defaultdict(lambda: 0)
+        self.total = collections.defaultdict(lambda: 0)
```
Comment (jscuds) [inline metrics.py:34]: Is the `lambda` function here used to instantiate an empty `defaultdict()`?

File: metrics.py [hunk 8]
```diff
@@ -1,27 +1,89 @@
...
+        
+        Returns: Dict that looks like
+        {
+            'metric_key': metric (float torch.Tensor): the average metric of shape ()
+        }
+        """
+        raise NotImplementedError()
+
+
+glob_n=0
```
Comment (jscuds) [inline metrics.py]: Where else is `glob_n` referenced?

File: train.py [hunk 9]
```diff
@@ -178,16 +182,17 @@ def run_training_loop(args: argparse.Namespace) -> str:
     # watch_log_freq is setup to log every 10 batches: num_examples//batch_size//10
     #    which means it will log gradients every `watch_log_freq` batches
 
-    watch_log_freq = args.num_examples//args.batch_size//10
+    watch_log_freq = round(len(train_dataloader) / args.batch_size / 10.0)
```
Comment (jscuds) [inline train.py]: I think for this to be the same...just remove `/ args.batch_size` so it reads:

> `round(len(train_dataloader) / 10.0)`

because `len(train_dataloader) == np.ceil(args.num_examples/args.batch_size)`

**Review by jscuds — COMMENTED** [review]
I have a few questions throughout.  Looks great!

File: .github/workflows/test.yml (hunk 1 above)
Comment (jxmorris12) [reply .github/workflows/test.yml:18]: yeah sounds good, that seems like a better way to do things vs. requiring python 3.9 like I wanted! thanks

File: experiment.py (hunk 2 above)
Comment (jxmorris12) [reply experiment.py:71]: good call 👍

File: train.py (hunk 3 above)
Comment (jxmorris12) [reply train.py:126]: yeah agree, You could do this with ** or `config_dict = vars(args)` and then `config_dict.update` as in your second suggestion

File: utils.py (hunk 4 above)
Comment (jxmorris12) [reply utils.py:55]: yeah sorry if this doesn't work quite right haha but we can always change it back since we don't need the histograms that badly right now

File: experiment.py (hunk 5 above)
Comment (jxmorris12) [reply experiment.py:234]: awesome I agree

File: metrics.py (hunk 7 above)
Comment (jxmorris12) [reply metrics.py:34]: yep, with the default argument of zero. 

so this is what it fixes...

old pattern:
```python
d = {}

for key in ['train', 'test']:
  if key not in d:
    d[key] = 0
  d[key] += 1
```
new pattern:
```
d = collections.defaultdict(lambda: 0)
for key in ['train', 'test']:
  d[key] += 1
```

hope that makes sense. btw I remembered `collections.Counter` could do the same thing as these but whatever.

File: metrics.py (hunk 8 above)
Comment (jxmorris12) [reply metrics.py]: sorry I need to remove this - was using it for debugging

File: train.py (hunk 9 above)
Comment (jxmorris12) [reply train.py]: Ok I realized I forgot that the dataloader length is in batches and --num_examples was in examples, so thanks for catching that bug. On the other hand though it's important that `args.num_examples` can be None now so it's not always equal to what you said

Comment (jscuds) [reply train.py]: Right!  I wasn't clear: I just meant "generic length of dataset."

File: experiment.py (hunk 6 above)
Comment (jxmorris12) [reply experiment.py:284]: yeah exactly. And metrics reset when `compute()` is called
